### PR TITLE
Fix for the non-determinism issue on Command handler and Outgoing Commands

### LINKED ISSUE #7135: tv-app gen/ files are modified by darwin helper.js changes
#### Problem
When adding OtaSoftwareUpdateServer Cluster to `src/darwin/Framework/CHIP/templates/helper.js` inside `function asExpectedEndpointForCluster(clusterName)`, files inside `tv-app/tv-common/gen/` are modified after running `zap_regen_all.py`.

Result: `tv-app` build fails

##### Steps to Reproduce:
1. checkout master
2. modify `helper.js` as such:
```
@@ -38,6 +38,7 @@ function asExpectedEndpointForCluster(clusterName)
   case 'NetworkCommissioning':
   case 'OperationalCredentials':
   case 'TrustedRootCertificates':
+  case 'OtaSoftwareUpdateServer':
     return 0;
   }
```
3. run `zap_regen_all.py`
4. Observe that `tv-app/` files have been modified. Observe that tv-app build fails

 #### Proposed Solution
Not sure. But it seems to me that changes to darwin `helper.js` should not affect `tv-app`.

### PR DESCRIPTION
**Pull request by brdandu**
- Adding changes to the zap templates such that the incoming and outgoing commands are generated with determinism. Using the upto date helpers in the *.zapt templates

Currently pointing to a zap repo which has new helpers as well. These changes have not been merged into the github zap repo
gihub#342

#### Problem
What is being fixed?  
 - Github issue [#342](https://github.com/project-chip/zap/issues/342)
 - This should fix the regen diff across regens
 - Fixes the generation for incoming and outgoing command handling

#### Change overview
Changes to ZAP generation templates: CHIPClusters.*zapt, im-cluster-command-handler.zapt to generate code accurately.
This PR depends on the following ZAP PR:
https://github.com/project-chip/zap/pull/388

#### Testing
How was this tested? (at least one bullet point required)
Regened multiple times to check that the CHIPClusters and im-cluster-command-handler generated files did not change across regens.

### CODE REVIEW

**Comment by brdandu** [comment]
This PR requires the helpers added in https://github.com/project-chip/zap/pull/388 PR

File: src/app/zap-templates/templates/app/CHIPClusters.zapt [hunk 1]
```diff
@@ -1,6 +1,7 @@
 {{> header}}
 
-{{#if (chip_has_client_clusters)}}
+{{#all_user_clusters_with_outgoing_commands}}
```
Comment (vivien-apple) [inline src/app/zap-templates/templates/app/CHIPClusters.zapt:3]: This is different from `chip_client_clusters` right ?  Looking at the output, it emit codes for the responses commands.

File: src/app/zap-templates/templates/app/CHIPClusters.zapt [hunk 2]
```diff
@@ -14,22 +15,23 @@
...
+{{/first}}
+class DLL_EXPORT {{asUpperCamelCase clusterName}}Cluster : public ClusterBase
 {
 public:
-    {{asUpperCamelCase name}}Cluster() : ClusterBase(app::Clusters::{{asUpperCamelCase name}}::Id) {}
-    ~{{asUpperCamelCase name}}Cluster() {}
-    {{#chip_cluster_commands}}
+    {{asUpperCamelCase clusterName}}Cluster() : ClusterBase(app::Clusters::{{asUpperCamelCase clusterName}}::Id) {}
+    ~{{asUpperCamelCase clusterName}}Cluster() {}
+    {{#all_outgoing_commands_for_cluster clusterName clusterSide isMfgSpecific="false"}}
```
Comment (vivien-apple) [inline src/app/zap-templates/templates/app/CHIPClusters.zapt:24]: Why do we use "isMfgSpecific=false" ? It removes "MfgSpecificPing". At some point we want to support manufacturer specific commands.

File: src/app/zap-templates/templates/app/CHIPClusters.zapt [hunk 3]
```diff
@@ -14,22 +15,23 @@
...
     {{#first}}
 
     // Cluster Commands
     {{/first}}
-    CHIP_ERROR {{asUpperCamelCase name}}(Callback::Cancelable * onSuccessCallback, Callback::Cancelable * onFailureCallback{{#chip_cluster_command_arguments_with_structs_expanded}}{{#if_is_struct type}}{{else}}, {{chipType}} {{asLowerCamelCase label}}{{/if_is_struct}}{{/chip_cluster_command_arguments_with_structs_expanded}});
-    {{/chip_cluster_commands}}
+    CHIP_ERROR {{asUpperCamelCase commandName}}(Callback::Cancelable * onSuccessCallback, Callback::Cancelable * onFailureCallback{{#zcl_command_arguments}}{{#if_is_struct type}}{{else}}, {{as_underlying_zcl_type type}} {{asLowerCamelCase label}}{{/if_is_struct}}{{/zcl_command_arguments}});
```
Comment (vivien-apple) [inline src/app/zap-templates/templates/app/CHIPClusters.zapt:29]: I have not looked the source code for {{zcl_command_arguments}} but I suspect that something is different compare to {{chip_cluster_command_arguments_with_structs_expanded}) since it seems that the command arguments have disappeared for the "controller-clusters.zap" configuration.

File: zzz_generated/tv-casting-app/zap-generated/IMClusterCommandHandler.cpp [hunk 4]
```diff
@@ -1309,15 +1309,6 @@ void DispatchServerCommand(CommandHandler * apCommandObj, const ConcreteCommandP
...
         }
-        case Commands::TestSpecific::Id: {
-            Commands::TestSpecific::DecodableType commandData;
-            TLVError = DataModel::Decode(aDataTlv, commandData);
-            if (TLVError == CHIP_NO_ERROR)
-            {
-                wasHandled = emberAfTestClusterClusterTestSpecificCallback(apCommandObj, aCommandPath, commandData);
-            }
-            break;
-        }
```
Comment (vivien-apple) [inline zzz_generated/tv-casting-app/zap-generated/IMClusterCommandHandler.cpp:1320]: I wonder why this is removed ?

**Review by vivien-apple — CHANGES_REQUESTED** [review]
It seems like the contents of `CHIPClusters.*` files are changed to contains codegen for command responses, which is not the intent there.

I'm also curious to understand why `TestSpecific` has dissapeared from the generated `I'm-cluster-command-handler` file.

File: src/app/zap-templates/templates/app/CHIPClusters.zapt (hunk 1 above)
Comment (brdandu) [reply src/app/zap-templates/templates/app/CHIPClusters.zapt:3]: Yes all_user_clusters_with_outgoing_commands is different from chip_has_client_clusters. I felt like we needed this for all the outgoing commands i.e. outgoing commands from client clusters as well as outgoing commands from server clusters. Correct me if I am wrong but the way I understood it is that the generated code was not handling outgoing commands from server to client. You are right that many of the server to client outgoing commands can be command responses but may be not all of them. Is there a reason we do not want this for command responses? Zigbee pro used to have something similar in the form of client-command-macro.h which used to generate for all outgoing commands disregarding the fact on whether the outgoing command was enabled or not by the user. This generated file seemed similar to that and based on the current strategy of just generating for enabled outgoing commands I added new helpers such that commands which are not enabled will not be generated for.

File: src/app/zap-templates/templates/app/CHIPClusters.zapt (hunk 2 above)
Comment (brdandu) [reply src/app/zap-templates/templates/app/CHIPClusters.zapt:24]: I could be wrong but I remember reading somewhere that mfg specific codes were not specifically used to denote that a cluster or command is mfg specific but the range of the cluster and command code is supposed to tell if it is mfg specific. Based on that assumption I turned this off since there will be no cluster or command code collisions with that assumption. However that is not the case for Zigbee 3.0

File: src/app/zap-templates/templates/app/CHIPClusters.zapt (hunk 3 above)
Comment (brdandu) [reply src/app/zap-templates/templates/app/CHIPClusters.zapt:29]: Once I changed the parent block helpers I could no longer use {{chip_cluster_command_arguments_with_structs_expanded}}. Would it be possible to remove the restrictions in chip_cluster_command_arguments_with_structs_expanded. It makes more sense to continue using this and not zcl_command_arguments. I also noticed that the zap helper files are now being created outside the zap repo, could we bring these back into the zap repo? I feel it makes more sense to have them in the zap repo.

File: zzz_generated/tv-casting-app/zap-generated/IMClusterCommandHandler.cpp (hunk 4 above)
Comment (brdandu) [reply zzz_generated/tv-casting-app/zap-generated/IMClusterCommandHandler.cpp:1320]: It seems correct that this was removed because I see that the test cluster server is only enabled in EP1. When I go to the command settings there I see that the incoming for "TestSpecific" command is disabled so it should not have been generated for before. The old helpers for incoming commands must have been outdated and were generating randomly. It should not do that with the new template changes in the commandHandler template. However I have to warn you that there is still one use case which has not been dealt with yet where you may have different command or attribute settings for the same cluster in different endpoints for eg the attribute default value. That is a complex situation and needs to be looked at in terms of the generated code and embedded code changes. I do not think we have hit that situation yet but still a possibility that a user may hit this.

**Comment by brdandu** [comment]
> It seems like the contents of `CHIPClusters.*` files are changed to contains codegen for command responses, which is not the intent there.
> 
> I'm also curious to understand why `TestSpecific` has dissapeared from the generated `I'm-cluster-command-handler` file.

I have responded to this in some form in the below comments as well. Any reason we do not have code gen for command responses? Also, are we sure that we are not excluding any outgoing commands from server to client which are not command responses?

**Review by andy31415 — CHANGES_REQUESTED** [review]
Please cleanup the patch summary to say what the problem was, what the fix was and how it was tested in the standard format.

Currently the template was kept, so it says "Frobnozzle is leaky" and such :)

**Review by andy31415 — DISMISSED** [review]

File: src/app/zap-templates/templates/app/CHIPClusters.zapt (hunk 1 above)
Comment (bzbarsky-apple) [reply src/app/zap-templates/templates/app/CHIPClusters.zapt:3]: There are no outgoing commands from server to client that we need codegen like this for.  All commands from server to client are responses in Matter.

In general, I would expect this PR to lead to no generated code changes except the cases where the current codegen is wrong (lighting-app, in particular).

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]

**Review by andy31415 — CHANGES_REQUESTED** [review]
This PR seems to change generated output. For a "Remove non-determinism" I would expect just that: do not alter output, just make it deterministic.

Why do the generated files change? If the change is needed, we should split PRs into 'change output' and 'remove determinism'

**Review by bzbarsky-apple — COMMENTED** [review]
This is getting much closer, and the lighting-app bugginness is fixed with this, which is nice!

**Comment by brdandu** [comment]
> This PR seems to change generated output. For a "Remove non-determinism" I would expect just that: do not alter output, just make it deterministic.
> 
> Why do the generated files change? If the change is needed, we should split PRs into 'change output' and 'remove determinism'

Got it cleaning it up based on determinism

File: zzz_generated/controller-clusters/zap-generated/CHIPClusters.h [hunk 5]
```diff
@@ -30,13 +30,6 @@
 namespace chip {
 namespace Controller {
 
-class DLL_EXPORT AccessControlCluster : public ClusterBase
```
[the remark this answers is not in the document]
Comment (brdandu) [reply zzz_generated/controller-clusters/zap-generated/CHIPClusters.h:33]: These clusters do not have any outgoing commands enabled. Why do we want these even when there are no outgoing commands from client to server?

File: zzz_generated/controller-clusters/zap-generated/CHIPClusters.cpp [hunk 6]
```diff
@@ -3187,24 +3086,6 @@ CHIP_ERROR GroupKeyManagementCluster::KeySetWrite(Callback::Cancelable * onSucce
...
-    // epochStartTime0: int64u
-    SuccessOrExit(err = writer->Put(TLV::ContextTag(argSeqNumber++), epochStartTime0));
-    // epochKey1: octetString
-    SuccessOrExit(err = writer->Put(TLV::ContextTag(argSeqNumber++), epochKey1));
-    // epochStartTime1: int64u
-    SuccessOrExit(err = writer->Put(TLV::ContextTag(argSeqNumber++), epochStartTime1));
-    // epochKey2: octetString
-    SuccessOrExit(err = writer->Put(TLV::ContextTag(argSeqNumber++), epochKey2));
-    // epochStartTime2: int64u
-    SuccessOrExit(err = writer->Put(TLV::ContextTag(argSeqNumber++), epochStartTime2));
```
[the remark this answers is not in the document]
Comment (brdandu) [reply zzz_generated/controller-clusters/zap-generated/CHIPClusters.cpp]: This is fixed now.

File: src/app/zap-templates/templates/app/CHIPClusters.zapt (hunk 2 above)
Comment (brdandu) [reply src/app/zap-templates/templates/app/CHIPClusters.zapt]: Removed the isMfgSpecific flag. The MfgSpecificPing is being generated now

**Review by andy31415 — APPROVED** [review]

**Review by vivien-apple — APPROVED** [review]

**Comment by vivien-apple** [comment]
R+ assuming Boris's comment is answered.

File: examples/lighting-app/lighting-common/lighting-app.matter [hunk 7]
```diff
@@ -957,6 +957,24 @@ client cluster OnOff = 6 {
   attribute enum8 startUpOnOff = 16387;
   readonly global attribute bitmap32 featureMap = 65532;
   readonly global attribute int16u clusterRevision = 65533;
+
+  request struct OffWithEffectRequest {
```
Comment (bzbarsky-apple) [inline examples/lighting-app/lighting-common/lighting-app.matter:961]: This is remaining non-determinism, right?  I assume this will be fixed in a followup?

File: zzz_generated/controller-clusters/zap-generated/CHIPClusters.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (bzbarsky-apple) [reply zzz_generated/controller-clusters/zap-generated/CHIPClusters.h:33]: Because as I said: "these structs are also used for attribute reads/writes"

**Review by bzbarsky-apple — COMMENTED** [review]
The issue with the tree not compiling obviously still needs to be addressed.

**Review by jmartinez-silabs — APPROVED** [review]

File: examples/lighting-app/lighting-common/lighting-app.matter (hunk 7 above)
Comment (brdandu) [reply examples/lighting-app/lighting-common/lighting-app.matter:961]: I have no idea what this file is and why this changed. @tecimovic do you know what this file is and why this is changing?

File: src/app/zap-templates/templates/app/CHIPClusters-src.zapt [hunk 8]
```diff
@@ -18,11 +18,13 @@ namespace Controller {
 // TODO(#4503): length should be passed to commands when byte string is in argument list.
 // TODO(#4503): Commands should take group id as an argument.
 
-{{#chip_client_clusters}}
-
-// {{asUpperCamelCase name}} Cluster Commands
-{{#chip_cluster_commands}}
-CHIP_ERROR {{asUpperCamelCase clusterName}}Cluster::{{asUpperCamelCase name}}(Callback::Cancelable * onSuccessCallback, Callback::Cancelable * onFailureCallback{{#chip_cluster_command_arguments_with_structs_expanded}}{{#if_is_struct type}}{{else}}, {{chipType}} {{asLowerCamelCase label}}{{/if_is_struct}}{{/chip_cluster_command_arguments_with_structs_expanded}})
+{{#all_user_clusters_with_outgoing_commands}}
```
Comment (vivien-apple) [inline src/app/zap-templates/templates/app/CHIPClusters-src.zapt:21]: Now that `CHIPClusters.zapt` uses back `chip_client_clusters` can we use it here too for consistency and couple it with `{{#all_outgoing_commands_for_cluster}}` ?

Comment (brdandu) [reply src/app/zap-templates/templates/app/CHIPClusters-src.zapt:21]: I can change it if you want but we get those 1 line comments with no corresponding code when there are no outgoing commands from client which seems confusing. If we still want it then I can make a change for that.

File: examples/lighting-app/lighting-common/lighting-app.matter (hunk 7 above)
Comment (bzbarsky-apple) [reply examples/lighting-app/lighting-common/lighting-app.matter:961]: This file is generated by ZAP, and presumably it's also using the helpers that are non-deterministic and its template needs to be fixed accordingly.  That can happen in a separate PR.

File: src/app/zap-templates/templates/app/CHIPClusters-src.zapt (hunk 8 above)
Comment (bzbarsky-apple) [reply src/app/zap-templates/templates/app/CHIPClusters-src.zapt:21]: If we put the comments inside a `{{#first}}` block we would only get them when we have something to show, no?

Comment (brdandu) [reply src/app/zap-templates/templates/app/CHIPClusters-src.zapt:21]: Makes sense. yes that is correct. Will switch the cluster loop back

**Review by bzbarsky-apple — APPROVED** [review]
Looks much better, thank you!  Vivien's question about the header and .cpp file potentially being out of sync due to using different iterators is the only issue I see remaining.

File: src/app/zap-templates/templates/app/CHIPClusters-src.zapt (hunk 8 above)
Comment (brdandu) [reply src/app/zap-templates/templates/app/CHIPClusters-src.zapt]: Using chip_client_clusters now

File: src/app/zap-templates/common/ClustersHelper.js [hunk 9]
```diff
@@ -649,9 +649,8 @@ Clusters.init = async function(context) {
 //
 // Helpers: All
 //
-function asBlocks(promise, options)
-{
-  return promise.then(data => templateUtil.collectBlocks(data, options, this))
+function asBlocks(promise, options) {
+  return promise.then(data => data ? templateUtil.collectBlocks(data, options, this) : '')
```
Comment (bzbarsky-apple) [inline src/app/zap-templates/common/ClustersHelper.js:653]: Please undo this change, and fix the caller in `chip_cluster_command_arguments_with_structs_expanded` to pass in an empty array as needed.

**Review by andy31415 — DISMISSED** [review]
I believe this patch may have picked unwanted third party repo changes.

Efr32 I am quite certain it is not needed.

ZAP was also rolled forward recently, so guessing that may also not be needed? please check.

**Review by andy31415 — APPROVED** [review]

File: examples/light-switch-app/light-switch-common/light-switch-app.matter [hunk 10]
```diff
@@ -580,6 +580,28 @@ client cluster Groups = 4 {
   readonly attribute bitmap8 nameSupport = 0;
   readonly global attribute attrib_id attributeList[] = 65531;
   readonly global attribute int16u clusterRevision = 65533;
+
+  request struct AddGroupRequest {
```
Comment (andy31415) [inline examples/light-switch-app/light-switch-common/light-switch-app.matter:584]: These requests are not used by any commands. Why are they generated?

**Comment by bzbarsky-apple** [comment]
Note that at this point https://github.com/project-chip/connectedhomeip/pull/14581 will remove some (but not all!) of the bits being changed in this PR....

File: examples/light-switch-app/light-switch-common/light-switch-app.matter (hunk 10 above)
Comment (brdandu) [reply examples/light-switch-app/light-switch-common/light-switch-app.matter:584]: I updated the responses which added the following
command AddGroup(AddGroupRequest): AddGroupResponse = 0;

Were you looking for that?

**Review by andy31415 — APPROVED** [review]

**Comment by brdandu** [comment]
> Note that at this point #14581 will remove some (but not all!) of the bits being changed in this PR....

Seems like the CHIPCluster and CHIPClientCallback templates are going away. Those are still non deterministic. Should I continue looking into fixing that?

File: examples/light-switch-app/light-switch-common/light-switch-app.matter (hunk 10 above)
Comment (bzbarsky-apple) [reply examples/light-switch-app/light-switch-common/light-switch-app.matter:584]: CHIPClusters and CHIPClientCallback no longer generate anything at all for commands on tip.  Please update to tip.....

File: examples/all-clusters-app/all-clusters-common/all-clusters-app.matter [hunk 11]
```diff
@@ -1647,23 +1647,23 @@ server cluster IasZone = 1280 {
...
+    INT8U zoneId = 1;
+  }
+
+  response struct ZoneEnrollRequest {
     IasZoneType zoneType = 0;
     INT16U manufacturerCode = 1;
   }
 
-  request struct ZoneStatusChangeNotificationRequest {
+  response struct ZoneStatusChangeNotification {
```
Comment (andy31415) [inline examples/all-clusters-app/all-clusters-common/all-clusters-app.matter:1660]: We seem to have removed ZoneEndrollResponse, however that response is used in the command ZoneEnrollREquest.

Also the name of the command argument is 'ZoneEnrollRequestRequest' and above the struct is defined with the wrong name.
